File: workers/extraction-api.py
"""
TopCinema Video Extraction API
Cloudflare Workers endpoint for direct video URL extraction
"""

import json
from typing import Dict, Optional

# Server extraction patterns
SERVER_PATTERNS = {
    'StreamWish': {
        'patterns': [
            r'file:"([^"]+\.m3u8[^"]*)"',
            r'source:\s*"([^"]+\.m3u8[^"]*)"'
        ]
    },
    'Mixdrop': {
        'patterns': [
            r'wurl="([^"]+)"',
            r'MDCore\.wurl="([^"]+)"'
        ],
        'post_process': lambda url: f"https:{url}" if not url.startswith('http') else url
    },
    'LuluStream': {
        'patterns': [
            r'file:"([^"]+\.m3u8[^"]*)"'
        ]
    }
}

async def fetch_server_iframe(movie_id: str, server_index: int) -> Optional[str]:
    """Fetch iframe URL from TopCinema API"""
    url = "https://topcinemaa.co/wp-content/themes/movies2023/Ajaxat/Single/Server.php"
    
    response = await fetch(url, {
        'method': 'POST',
        'headers': {
            'Content-Type': 'application/x-www-form-urlencoded',
            'X-Requested-With': 'XMLHttpRequest'
        },
        'body': f'id={movie_id}&i={server_index}'
    })
    
    html = await response.text()
    
    # Extract iframe src
    import re
    match = re.search(r'src="([^"]+)"', html)
    return match.group(1) if match else None


async def extract_video_url(iframe_url: str, server_name: str) -> Optional[str]:
    """Extract direct video URL from server iframe"""
    try:
        response = await fetch(iframe_url, {
            'headers': {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
                'Referer': 'https://topcinemaa.co/'
            }
        })
        
        html = await response.text()
        
        if server_name not in SERVER_PATTERNS:
            return None
        
        config = SERVER_PATTERNS[server_name]
        
        for pattern in config['patterns']:
            import re
            matches = re.findall(pattern, html)
            if matches:
                url = matches[0]
                
                # Apply post-processing if exists
                if 'post_process' in config:
                    url = config['post_process'](url)
                
                return url
        
        return None
        
    except Exception as e:
        print(f"Extraction error: {e}")
        return None


async def handle_request(request):
    """Main request handler for Cloudflare Workers"""
    
    # CORS headers
    cors_headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Content-Type': 'application/json'
    }
    
    # Handle OPTIONS
    if request.method == 'OPTIONS':
        return Response(None, {'status': 204, 'headers': cors_headers})
    
    try:
        url = URL(request.url)
        path = url.pathname
        
        # GET /api/servers/:movieId - Get all server URLs
        if path.startswith('/api/servers/'):
            movie_id = path.split('/')[-1]
            
            servers = []
            server_names = [
                'VideoTube', 'UpDown', 'StreamWish', 'Doodstream',
                'Filelions', 'Streamtape', 'LuluStream', 'Mixdrop'
            ]
            
            for i, name in enumerate(server_names):
                iframe_url = await fetch_server_iframe(movie_id, i)
                if iframe_url:
                    servers.append({
                        'id': i,
                        'name': name,
                        'iframe_url': iframe_url,
                        'extractable': name in SERVER_PATTERNS
                    })
            
            return Response(
                json.dumps({'success': True, 'servers': servers}),
                {'headers': cors_headers}
            )
        
        # POST /api/extract - Extract video URL
        elif path == '/api/extract':
            data = await request.json()
            iframe_url = data.get('iframe_url')
            server_name = data.get('server_name')
            
            if not iframe_url or not server_name:
                return Response(
                    json.dumps({'success': False, 'error': 'Missing parameters'}),
                    {'status': 400, 'headers': cors_headers}
                )
            
            video_url = await extract_video_url(iframe_url, server_name)
            
            if video_url:
                return Response(
                    json.dumps({'success': True, 'video_url': video_url}),
                    {'headers': cors_headers}
                )
            else:
                return Response(
                    json.dumps({'success': False, 'error': 'Could not extract video URL'}),
                    {'status': 404, 'headers': cors_headers}
                )
        
        # GET /api/extract/:movieId/:server - Direct extraction
        elif path.startswith('/api/extract/'):
            parts = path.split('/')
            if len(parts) >= 5:
                movie_id = parts[3]
                server_index = int(parts[4])
                
                server_names = [
                    'VideoTube', 'UpDown', 'StreamWish', 'Doodstream',
                    'Filelions', 'Streamtape', 'LuluStream', 'Mixdrop'
                ]
                
                if server_index >= len(server_names):
                    return Response(
                        json.dumps({'success': False, 'error': 'Invalid server index'}),
                        {'status': 400, 'headers': cors_headers}
                    )
                
                server_name = server_names[server_index]
                iframe_url = await fetch_server_iframe(movie_id, server_index)
                
                if not iframe_url:
                    return Response(
                        json.dumps({'success': False, 'error': 'Could not fetch iframe URL'}),
                        {'status': 404, 'headers': cors_headers}
                    )
                
                video_url = await extract_video_url(iframe_url, server_name)
                
                if video_url:
                    return Response(
                        json.dumps({
                            'success': True,
                            'server': server_name,
                            'iframe_url': iframe_url,
                            'video_url': video_url
                        }),
                        {'headers': cors_headers}
                    )
                else:
                    return Response(
                        json.dumps({
                            'success': False,
                            'server': server_name,
                            'iframe_url': iframe_url,
                            'error': 'Could not extract video URL'
                        }),
                        {'status': 404, 'headers': cors_headers}
                    )
        
        # Default response
        return Response(
            json.dumps({
                'name': 'TopCinema Video Extraction API',
                'version': '1.0.0',
                'endpoints': {
                    'GET /api/servers/:movieId': 'Get all server URLs for a movie',
                    'POST /api/extract': 'Extract video URL from iframe',
                    'GET /api/extract/:movieId/:serverIndex': 'Direct extraction'
                }
            }),
            {'headers': cors_headers}
        )
        
    except Exception as e:
        return Response(
            json.dumps({'success': False, 'error': str(e)}),
            {'status': 500, 'headers': cors_headers}
        )


# Cloudflare Workers entry point
addEventListener('fetch', event => {
    event.respondWith(handle_request(event.request))
})

File: workers/extraction-api.js
/**
 * TopCinema Video Extraction API
 * Cloudflare Workers endpoint for direct video URL extraction
 */

// Server extraction patterns
const SERVER_PATTERNS = {
  'StreamWish': {
    patterns: [
      /file:"([^"]+\.m3u8[^"]*)"/,
      /source:\s*"([^"]+\.m3u8[^"]*)"/
    ]
  },
  'Mixdrop': {
    patterns: [
      /wurl="([^"]+)"/,
      /MDCore\.wurl="([^"]+)"/
    ],
    postProcess: (url) => url.startsWith('http') ? url : `https:${url}`
  },
  'LuluStream': {
    patterns: [
      /file:"([^"]+\.m3u8[^"]*)"/
    ]
  }
};

/**
 * Fetch iframe URL from TopCinema API
 */
async function fetchServerIframe(movieId, serverIndex) {
  const url = 'https://topcinemaa.co/wp-content/themes/movies2023/Ajaxat/Single/Server.php';
  
  const formData = new URLSearchParams();
  formData.append('id', movieId);
  formData.append('i', serverIndex);
  
  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'X-Requested-With': 'XMLHttpRequest'
      },
      body: formData.toString()
    });
    
    const html = await response.text();
    const match = html.match(/src="([^"]+)"/);
    return match ? match[1] : null;
  } catch (error) {
    console.error('Error fetching iframe:', error);
    return null;
  }
}

/**
 * Extract direct video URL from server iframe
 */
async function extractVideoUrl(iframeUrl, serverName) {
  try {
    const response = await fetch(iframeUrl, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Referer': 'https://topcinemaa.co/'
      }
    });
    
    const html = await response.text();
    
    if (!SERVER_PATTERNS[serverName]) {
      return null;
    }
    
    const config = SERVER_PATTERNS[serverName];
    
    for (const pattern of config.patterns) {
      const match = html.match(pattern);
      if (match) {
        let url = match[1];
        
        // Apply post-processing if exists
        if (config.postProcess) {
          url = config.postProcess(url);
        }
        
        return url;
      }
    }
    
    return null;
    
  } catch (error) {
    console.error('Extraction error:', error);
    return null;
  }
}

/**
 * CORS headers
 */
const corsHeaders = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type',
  'Content-Type': 'application/json'
};

/**
 * Main request handler
 */
async function handleRequest(request) {
  const url = new URL(request.url);
  const path = url.pathname;
  
  // Handle OPTIONS
  if (request.method === 'OPTIONS') {
    return new Response(null, {
      status: 204,
      headers: corsHeaders
    });
  }
  
  try {
    // GET /api/servers/:movieId - Get all server URLs
    if (path.startsWith('/api/servers/')) {
      const movieId = path.split('/').pop();
      
      const serverNames = [
        'VideoTube', 'UpDown', 'StreamWish', 'Doodstream',
        'Filelions', 'Streamtape', 'LuluStream', 'Mixdrop'
      ];
      
      const servers = [];
      
      for (let i = 0; i < serverNames.length; i++) {
        const iframeUrl = await fetchServerIframe(movieId, i);
        if (iframeUrl) {
          servers.push({
            id: i,
            name: serverNames[i],
            iframe_url: iframeUrl,
            extractable: SERVER_PATTERNS[serverNames[i]] !== undefined
          });
        }
      }
      
      return new Response(
        JSON.stringify({ success: true, servers }),
        { headers: corsHeaders }
      );
    }
    
    // POST /api/extract - Extract video URL
    if (path === '/api/extract' && request.method === 'POST') {
      const data = await request.json();
      const { iframe_url, server_name } = data;
      
      if (!iframe_url || !server_name) {
        return new Response(
          JSON.stringify({ success: false, error: 'Missing parameters' }),
          { status: 400, headers: corsHeaders }
        );
      }
      
      const videoUrl = await extractVideoUrl(iframe_url, server_name);
      
      if (videoUrl) {
        return new Response(
          JSON.stringify({ success: true, video_url: videoUrl }),
          { headers: corsHeaders }
        );
      } else {
        return new Response(
          JSON.stringify({ success: false, error: 'Could not extract video URL' }),
          { status: 404, headers: corsHeaders }
        );
      }
    }
    
    // GET /api/extract/:movieId/:serverIndex - Direct extraction
    if (path.startsWith('/api/extract/')) {
      const parts = path.split('/').filter(p => p);
      
      if (parts.length >= 4) {
        const movieId = parts[2];
        const serverIndex = parseInt(parts[3]);
        
        const serverNames = [
          'VideoTube', 'UpDown', 'StreamWish', 'Doodstream',
          'Filelions', 'Streamtape', 'LuluStream', 'Mixdrop'
        ];
        
        if (serverIndex >= serverNames.length) {
          return new Response(
            JSON.stringify({ success: false, error: 'Invalid server index' }),
            { status: 400, headers: corsHeaders }
          );
        }
        
        const serverName = serverNames[serverIndex];
        const iframeUrl = await fetchServerIframe(movieId, serverIndex);
        
        if (!iframeUrl) {
          return new Response(
            JSON.stringify({ success: false, error: 'Could not fetch iframe URL' }),
            { status: 404, headers: corsHeaders }
          );
        }
        
        const videoUrl = await extractVideoUrl(iframeUrl, serverName);
        
        if (videoUrl) {
          return new Response(
            JSON.stringify({
              success: true,
              server: serverName,
              iframe_url: iframeUrl,
              video_url: videoUrl
            }),
            { headers: corsHeaders }
          );
        } else {
          return new Response(
            JSON.stringify({
              success: false,
              server: serverName,
              iframe_url: iframeUrl,
              error: 'Could not extract video URL'
            }),
            { status: 404, headers: corsHeaders }
          );
        }
      }
    }
    
    // Default response - API documentation
    return new Response(
      JSON.stringify({
        name: 'TopCinema Video Extraction API',
        version: '1.0.0',
        endpoints: {
          'GET /api/servers/:movieId': 'Get all server URLs for a movie',
          'POST /api/extract': 'Extract video URL from iframe (body: {iframe_url, server_name})',
          'GET /api/extract/:movieId/:serverIndex': 'Direct extraction (0-7)'
        },
        example: {
          get_servers: '/api/servers/240823',
          extract: '/api/extract/240823/2'
        }
      }),
      { headers: corsHeaders }
    );
    
  } catch (error) {
    return new Response(
      JSON.stringify({ success: false, error: error.message }),
      { status: 500, headers: corsHeaders }
    );
  }
}

// Cloudflare Workers entry point
addEventListener('fetch', event => {
  event.respondWith(handleRequest(event.request));
});
